internal/kafka/internal/workers/cluster_mgrs/dynamic_scale_down_mgr.go

```diff
@@ -0,0 +1,321 @@
+var _ dynamicScaleDownExecutor = &standardDynamicScaleDownProcessor{}
+var _ dynamicScaleDownProcessor = &standardDynamicScaleDownProcessor{}
+
+// ShouldScaleDown indicates whether a new data plane
+// cluster should be created for a given instance type in the given provider
+// and region.
+// It returns true if all the following conditions happen:
+// 1. If specified the cluster is empty i.e it does not contain any streaming unit
+// 2. If the cluster can be removed without triggering a scale up action
+// Otherwise false is returned.
+// Note: This method assumes kafkaStreamingUnitCountPerClusterList does not
+// contain elements with the Status attribute with the 'failed' value.
+// Thus, if the type is constructed with the assumptions being true, it
+// can be considered as the 'failed' state Clusters are not included in
+// the calculations.
+func (p *standardDynamicScaleDownProcessor) ShouldScaleDown() (bool, error) {
+	// First let's check if the cluster is empty
+	for _, i := range p.indexesOfStreamingUnitForSameClusterID {
+		clusterIsNotEmpty := p.kafkaStreamingUnitCountPerClusterList[i].Count > 0
+		if clusterIsNotEmpty {
+			return false, nil
+		}
+	}
+
+	// let's check if the cluster can be safely removed without causing a scale up event
+	if len(p.regionsSupportedInstanceType) == 0 { // if no region limits are available it means that this cluster is in a region that's not supported anymore, we can safely delete it if it is empty
+		return true, nil
+	}
+
+	newkafkaStreamingUnitCountPerClusterList := services.KafkaStreamingUnitCountPerClusterList{}
+
+	clusterStatesTowardReadyState := []string{
+		api.ClusterProvisioning.String(), api.ClusterProvisioned.String(),
+		api.ClusterAccepted.String(), api.ClusterWaitingForKasFleetShardOperator.String(),
+	}
+
+	for _, suCount := range p.kafkaStreamingUnitCountPerClusterList {
+		// Ignore clusters in terraforming states for scale up evaluation as they are not ready yet.
+		// We only want to delete the cluster if we've a sibling ready cluster
+		if arrays.Contains(clusterStatesTowardReadyState, suCount.Status) {
+			continue
+		}
+
+		// Keep only clusters streaming unit count for cluster that are not going to be deleted.
+		// i.e Assume that the candidate cluster is removed from the new list which will be used to perform scale up evaluation
+		if suCount.ClusterId == p.clusterID {
+			continue
+		}
+
+		newkafkaStreamingUnitCountPerClusterList = append(newkafkaStreamingUnitCountPerClusterList, suCount)
+	}
+
+	for _, i := range p.indexesOfStreamingUnitForSameClusterID {
+		suCount := p.kafkaStreamingUnitCountPerClusterList[i]
+
+		currLocator := supportedInstanceTypeLocator{
+			provider:         suCount.CloudProvider,
+			region:           suCount.Region,
+			instanceTypeName: suCount.InstanceType,
+		}
+
+		instanceTypeConfig, ok := p.regionsSupportedInstanceType[suCount.InstanceType]
+		if !ok {
+			continue
+		}
+
+		dynamicScaleUpProcessor := &standardDynamicScaleUpProcessor{
```

Comment:
I think passing the scale up processor as a dependency of this type as an interface might be a good idea as it would let us unit test with custom dynamic scale up implementations

Reply:
For the standard scale down processor, I wanted to use the standard scale up processor i.e the same scale up logic.

Reply:
I think the comment was not understood.
What I am suggesting is that a `dynamicScaleUpProcessor` (an interface) is passed to the standardDynamicScaleDownProcessor` instead of instantiating it inside with a concrete type

Reply:
I think this might not be trivial without refactoring as the instantiation depends on new data calculated in the method

Reply:
Yes, that's what I understood as well and I do not think that (customization) is needed. The standard scale down processor needs the standard scale up processor concrete type (basically it needs to evaluate if scale up can happen based on the logic in the standard scale up processor and nothing else). 

As such, the unit test for the `standardDynamicScaleDownProcessor` have been written with this "ensuring no scale up will be triggered" logic in mind.